### PR TITLE
Check if all states are covered

### LINKED ISSUE #293: Check if all states are provided when using `reinit=false`
### Wflow version checks

- [X] I have checked that this issue has not already been reported.
- [X] I have checked that this bug exists on the latest version of Wflow.

### Reproducible Example

If you start a wflow_sbm model with `glacier=true`, and don't pass the glacierstore in the state section, there is no error (or warning thrown.

### Current behaviour

See above

### Desired behaviour

When using `reinit=false`, it might be worth that Wflow checks whether all expected states are provided by the instate file. If not, at least a warning should be thrown (and perhaps even an error). 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by JoostBuitink**
## Issue addressed
Fixes #293 

## Explanation
Explain how you addressed the bug/feature request, what choices you made and why.

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `master`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed
- [x] Updated changelog.md if needed

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

### CODE REVIEW

File: src/states.jl [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+"""
+    extract_required_states(config)
+
+Function to retrieve the required states, given the model configuration. The required states
+are inferred from the model settings (mainly model_type, model options and routing types).
+Returns as list of required states, in the same formats as the keys that are returned from the
+`ncnames` function.
+"""
+function extract_required_states(config)
```
Comment (verseve) [inline src/states.jl:90]:
```suggestion
"""
    extract_required_states(config::Config)

Function to retrieve the required states, given the model configuration. The required states
are inferred from the model settings (mainly model_type, model options and routing types).
Returns as list of required states, in the same formats as the keys that are returned from the
`ncnames` function.
"""
function extract_required_states(config::Config)
```

File: src/states.jl [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    state_list::Union{Nothing,Tuple},
+)
+
+    if !isnothing(state_list)
+        for state in state_list
+            required_states = (required_states..., (key_entry..., state))
+        end
+    end
+    return required_states
+end
```
Comment (verseve) [inline src/states.jl:80]:
```suggestion
"""
    add_to_required_states(required_states::Tuple, key_entry::Tuple, states::Union{Nothing,Tuple})

Function to iterate through the list of `states` and add these to the `required_states`
variable. This is a tuple of tuples, that contains the correct "prefix" to the state
category (provided by the `key_entry` variable). This is added in front of each element in
the list of `states`, and the `required_states` tuple is returned.
"""
function add_to_required_states(
    required_states::Tuple,
    key_entry::Tuple,
    states::Tuple,
)
    for state in states
        required_states = (required_states..., (key_entry..., state))
    end

    return required_states
end

add_to_required_states(required_states::Tuple, key_entry::Tuple, states::Nothing,) = required_states
```

I suggest to make use of dispatch in this case.

File: docs/src/changelog.md [hunk 3]
```diff
@@ -29,6 +29,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Total water storage as an export variable for `SBM` concept. This is the total water stored
   per grid cell in millimeters. Excluded from this variable are the floodplain, lakes and
   reservoirs.
+- Checks to see if all states are covered in the .toml. If not all states are covered, an error
+  is thrown. If there are more states then required specified, these states are ignored (with a
+  warning in the logging), and the simulation will continue.
```
Comment (verseve) [inline docs/src/changelog.md:34]:
```suggestion
- Checks to see if all states are covered in the .toml file. If not all states are covered,
  an error is thrown. If there are more states specified than required, these states are
  ignored (with a warning in the logging), and the simulation will continue.
```

**Review by verseve — CHANGES_REQUESTED** [review]
Looks overall good to me. Some code suggestions, mainly related to argument types, dispatch and couple of typos. Nice addition of extra tests for this functionality! 

Another general suggestion is to reformat the code after the review if multiple files without code changes (for the PR) are affected. This makes it easier to check what has been changed as part of the PR :smiley:.

**Comment by JoostBuitink** [comment]
Thanks for the review, and good suggestions! I have included them all in the latest version. And sorry for creating a messy PR with the formatting changes, I will pay attention to that next time ;)

**Review by verseve — APPROVED** [review]
LGTM!

**Pull request merged**